core/chains/evm/log/registrations.go

```diff
@@ -196,80 +240,83 @@ func filtersContainValues(topicValues []common.Hash, filters [][]Topic) bool {
+	if _, exists := r.lookupSubscribers[addr]; !exists {
+		r.lookupSubscribers[addr] = make(map[common.Hash]map[*subscriber][][]Topic)
 	}
 
 	for topic, topicValueFilters := range sub.opts.LogsWithTopics {
-		if _, exists := r.handlersByAddr[addr][topic]; !exists {
-			r.handlersByAddr[addr][topic] = make(map[Listener]*listenerMetadata)
+		if _, exists := r.lookupSubscribers[addr][topic]; !exists {
+			r.lookupSubscribers[addr][topic] = make(map[*subscriber][][]Topic)
 			needsResubscribe = true
 		}
 
-		r.handlersByAddr[addr][topic][sub.listener] = &listenerMetadata{
-			opts:    sub.opts,
-			filters: topicValueFilters,
-		}
+		r.lookupSubscribers[addr][topic][sub] = topicValueFilters
 	}
 	return
 }
 
-func (r *handlers) removeSubscriber(sub *subscriber) (needsResubscribe bool) {
+func (r *handler) removeSubscriber(sub *subscriber) (needsResubscribe bool) {
 	addr := sub.opts.Contract
 
-	// FIXME: What about the case where you remove/add a job with the same contract address?
-	// addr is not good enough to be a unique key
-	// MARK MARK MARK
-	if _, exists := r.handlersByAddr[addr]; !exists {
-		return
-	}
 	for topic := range sub.opts.LogsWithTopics {
-		topicMap, exists := r.handlersByAddr[addr][topic]
+		// OK to panic on missing addr/topic here, since that would be an invariant violation:
+		// Both addr and topic will always have been added on addSubscriber
+		// LogsWithTopics should never be mutated
+		// Only removeSubscriber should ever remove anything from this map
+		addrTopics, exists := r.lookupSubscribers[addr]
 		if !exists {
-			continue
+			r.logger.Panicf("AssumptionViolation: expected lookupSubscribers to contain addr %s for subscriber %p with job ID %v", addr.Hex(), sub, sub.listener.JobID())
 		}
+		topicMap, exists := addrTopics[topic]
+		if !exists {
+			r.logger.Panicf("AssumptionViolation: expected addrTopics to contain topic %v for subscriber %p with job ID %v", topic, sub, sub.listener.JobID())
+		}
+		if _, exists = topicMap[sub]; !exists {
+			r.logger.Panicf("AssumptionViolation: expected topicMap to contain subscriber %p with job ID %v", sub, sub.listener.JobID())
+		}
+		delete(topicMap, sub)
 
-		delete(topicMap, sub.listener)
-
+		// cleanup and resubscribe
 		if len(topicMap) == 0 {
 			needsResubscribe = true
-			delete(r.handlersByAddr[addr], topic)
+			delete(r.lookupSubscribers[addr], topic)
 		}
-		if len(r.handlersByAddr[addr]) == 0 {
-			delete(r.handlersByAddr, addr)
+		if len(r.lookupSubscribers[addr]) == 0 {
+			delete(r.lookupSubscribers, addr)
 		}
 	}
 	return
 }
 
-func (r *handlers) addressesAndTopics() ([]common.Address, []common.Hash) {
+func (r *handler) addressesAndTopics() ([]common.Address, []common.Hash) {
 	var addresses []common.Address
 	var topics []common.Hash
-	for addr := range r.handlersByAddr {
+	for addr := range r.lookupSubscribers {
 		addresses = append(addresses, addr)
-		for topic := range r.handlersByAddr[addr] {
+		for topic := range r.lookupSubscribers[addr] {
```

Comment:
Could do s/lookupSubscribers/lookupSubs to reduce verbosity all around? :shrug: 
```suggestion
		for topic := range r.lookupSubs[addr] {
```